tests/Model/BSONIteratorTest.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace MongoDB\Tests\Model;
+
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Model\BSONArray;
+use MongoDB\Model\BSONDocument;
+use MongoDB\Model\BSONIterator;
+use MongoDB\Tests\TestCase;
+use stdClass;
+
+class BSONIteratorTest extends TestCase
+{
+    /**
+     * @dataProvider provideTypeMapOptionsAndExpectedDocuments
+     */
+    public function testValidValues(array $typeMap, $binaryString, array $expectedDocuments)
+    {
+        $bsonIt = new BSONIterator($binaryString, $typeMap);
+
+        $this->assertSameDocuments($expectedDocuments, $bsonIt);
```

Comment:
I just remembered that `assertSameDocuments()` converts documents to JSON before it compares, so it's not useful for testing that type maps have been applied. See `AggregateFunctionalTest::testTypeMapOption()` and `provideTypeMapOptionsAndExpectedDocuments()` for a good approach.